js/client-chat.js

```diff
@@ -185,10 +185,17 @@
 		// highlight
 
 		getHighlight: function (message) {
-			var highlights = Tools.prefs('highlights') || [];
-			if (!app.highlightRegExp) {
+			var allHighlights = Tools.prefs('highlights') || {};
+			if (allHighlights.constructor === 'Array') this.convertHighlights();
+			var thisroom = (this.type === 'battle' ? 'battlehls' : this.id);
+			var serverHighlights = allHighlights[Config.server.id] || {};
+			if (!serverHighlights['globalhls']) serverHighlights['globalhls'] = [];
+			if (!serverHighlights[thisroom]) serverHighlights[thisroom] = [];
+			if (!app.highlightRegExp || !app.highlightRegExp['globalhls'] || !app.highlightRegExp[thisroom]) {
```

Comment:
I don't know if I like how complicated this highlight scheme is.

You can have a `serverid.roomid` key system without needing nested objects, for the record.